core/src/main/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/DiskBasedChunkCache.java

```diff
@@ -0,0 +1,119 @@
+    static final String CACHE_DIRECTORY = "cache";
+
+    private DiskBasedChunkCacheConfig config;
+
+    public DiskBasedChunkCache(final ChunkManager chunkManager) {
+        super(chunkManager);
+    }
+
+    @Override
+    public InputStream cachedChunkToInputStream(final Path cachedChunk) {
+        try {
+            return Files.newInputStream(cachedChunk);
+        } catch (final IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Writes the chunk to specified directory on disk.
+     * To be able to understand if the files are fully written to disk during a potential recovery process,
+     * this implementation initially writes them to temporary directory and then atomically copies them
+     * to the actual caching directory.
+     */
+    @Override
+    public Path cacheChunk(final ChunkKey chunkKey, final InputStream chunk) throws IOException {
+        final Path tempChunkPath = config.baseCachePath().resolve(TEMP_CACHE_DIRECTORY)
+                .resolve(chunkKey.uuid.toString())
+                .resolve(String.valueOf(chunkKey.chunkId));
+        if (Files.notExists(tempChunkPath.getParent())) {
+            Files.createDirectories(tempChunkPath.getParent());
+        }
+        final Path tempCached = writeToDisk(chunk, tempChunkPath);
+        log.info("Chunk file has been stored to temporary caching directory {}", tempCached);
+        final Path cachedChunkPath = config.baseCachePath()
+                .resolve(CACHE_DIRECTORY).resolve(chunkKey.uuid.toString())
+                .resolve(String.valueOf(chunkKey.chunkId));
+        if (Files.notExists(cachedChunkPath.getParent())) {
+            Files.createDirectories(cachedChunkPath.getParent());
+        }
+        final Path newPath = Files.move(tempCached, cachedChunkPath, ATOMIC_MOVE);
+        log.debug("Chunk file has been moved to cache directory {}", newPath);
+        return newPath;
+    }
+
+    private static Path writeToDisk(final InputStream chunk, final Path tempChunkPath) throws IOException {
+        try (chunk) {
+            return Files.write(tempChunkPath, chunk.readAllBytes());
+        }
+    }
+
+    @Override
+    public RemovalListener<ChunkKey, Path> removalListener() {
+        return (key, path, cause) -> {
+            try {
+                Files.delete(path);
+                log.debug("Deleted cached file for key {} with path {} from cache directory."
+                        + " The reason of the deletion is {}", key, path, cause);
+            } catch (final IOException e) {
+                log.error("Failed to delete cached file for key {} with path {} from cache directory."
+                              + " The reason of the deletion is {}", key, path, cause, e);
+            }
+        };
+    }
+
+    @Override
+    public Weigher<ChunkKey, Path> weigher() {
+        return (key, value) -> {
+            try {
+                return Long.valueOf(Files.size(value)).intValue();
```

Comment:
[Nit] I guess that's the same
```suggestion
                return (int) Files.size(value));
```

Reply:
Also, this is not safe regarding sizes over `Integer.MAX_VALUE`. Should we cap the weight to `Integer.MAX_VALUE`? (Despite this shouldn't ever happen, but just in case)

Reply:
How would you suggest to cap it? The only way that comes to my mind is throwing an exception that the chunk is too big. The purpose `Long.valueOf(Files.size(value)).inValue()` was to get rid of the warning about unsafe conversion from long to int and offload this to JDK despite the fact that this is exactly the same. But you might be right that it's better to just cast.

Reply:
> How would you suggest to cap it? The only way that comes to my mind is throwing an exception that the chunk is too big.

Maybe just return max_int if the value >= max_int (compared in the long space). It's a safety net for a not-sure-if-ever-happens case

Reply:
I will add a warn log message just in case. Otherwise it will be really hard to track if something like this will go wrong.

Reply:
Fixed

Reply:
Yeah, looks good!

Maybe `<=` would work as well.

Reply:
Yeah, fixed